db/dcrpg/pgblockchain.go

```diff
@@ -547,12 +555,32 @@ func (pgb *ChainDB) GetTicketPoolByDateAndInterval(maturityBlock int64,
 	return retrieveTicketsByDate(pgb.db, maturityBlock, int64(val))
 }
 
-// TicketPoolVisualization fetches the following ticketpool data: tickets
+// TicketPoolVisualization helps block consecutive access to the ticketPoolVisualization
+// method till the first request is resolved.
+func (pgb *ChainDB) TicketPoolVisualization(interval dbtypes.ChartGrouping) (
+	[]*dbtypes.PoolTicketsData, *dbtypes.PoolTicketsData, error) {
+
+	isLocking := pgb.tpSync.updating.TryLock()
+
+	if isLocking {
+		defer pgb.tpSync.updating.Unlock()
+		// check if after the first request handled, the consecutive request response
+		// data was also processed and stored in the cache.
+		barcharts, donutCharts, ok := explorer.GetTicketPoolData(interval)
+		if ok {
+			return barcharts, donutCharts, nil
+		}
+		return pgb.ticketPoolVisualization(interval)
+	}
+	return nil, nil, nil
```

Comment:
We're still missing the mark with these edits.  There should be no path that returns no data.

There is a data request function that ensures fresh data, and which calls an update function if the data is not fresh.

A **data request** call:

- Get cached data (with a RLock) if it is not stale.
- If the cache was fresh, RUnlock and be done.
- If the cached data was stale/absent, the caller should attempt to **update** the cache.  This is where things are tricky.

An **update request** call (remember, multiple callers may be following this path, and we do not want multiple duplicate queries *or* failed requests / empty responses):
- Callers should attempt to get the "OK to update" lock (the trylock) in case the cached data is stale.  This lock indicates if a query is already running.
- Depending on if the caller got the updating lock, there are different paths, but returning with no data is not one.
- Regardless of whether the caller got the updating lock or not, wait on the data lock.  The data lock is needed:
  - If the updating lock was not obtained, the data lock is needed because there is an update in progress and they must wait for the data in cache to be refreshed!
  - If the updating lock was obtained, the lock is needed because the caller is about to do the update themselves.
  - Either way, callers must wait for the update to complete (when the data lock is free).
- A caller that **did not get the updating lock** will eventually be unblocked when the updating caller finishes and releases the data lock.  At this point they will return from the update function (when the cache is finally updated).
- A caller that **did get the updating lock** will finish the actual query and cache update, release the updating lock *and* the data lock, and return indicating that the cache was updated.
- On returning to the data request function, which called the update function, fresh data may be obtained from the cache.

Reply:
At that point where no data is expected, I should actually return an error that something went wrong with the trylock. Because the all the ticketPoolVisualization requests are suppose to be pilled and resolved one at a time by checking the actual cache or making the db call if the data doesn't exist.

Reply:
`return nil, nil, nil` we don't expect it to ever execute since something will have gone wrong and therefore I should return an error that the trylock did not work as expected?

Reply:
Maybe the question to ask is, Is trylock a blocking operation when the resource is not available for locking?

Reply:
I am thinking then of a for loop and ticker will help here.